src/data-model-providers/codegen/ClusterIntegration.h

```diff
@@ -0,0 +1,105 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <app/server-cluster/ServerClusterInterfaceRegistry.h>
+
+#include <cstdint>
+
+namespace chip::app {
+
+/// Handles reusable code for integrating server cluster interfaces with ember:
+///   - loads ember metadata (optional attributes and feature maps)
+///   - calls corresponding register/unregister calls on the Codegen Data Model Provider.
+///
+/// Uses an underlying delegate to handle constructor/destructor calls for actual cluster implementations.
+class CodegenClusterIntegration
+{
+public:
+    /// Handles the creation/destruction of server clusters when using codegen integration.
+    ///
+    /// The purpose of this class is to call ServerClusterInterfaces constructors and
+    /// destructors once generic ember processing has been completed.
+    ///
+    /// Logic is generally split into:
+    ///   - generic fetching and validating data from the ember framework like
+    ///     0-based array index from endpoint id, optional attribute fetching, feature map fetching
+    ///   - registering/unregistering ServerClusterInterfaces
+    ///   - determining how to create/destroy classes (the purpose of this delegate)
+    class Delegate
+    {
+    public:
+        virtual ~Delegate() = default;
+
+        /// Create the given registration for an endpoint
+        ///
+        /// When this is called, the caller has ensured that the endpointId was valid, that optionalAttributeBits are loaded
+        /// from ember according to supported attributes and that the feature map for the underlying cluster ID has been loaded.
+        ///
+        /// NOTE: optionalAttributeBits is intended for low id attributes since it supports attribute bits up to 31 only. It is
+        ///       intended for use with `OptionalAttributeSet` and NOT all clusters support this. Specific examples:
+        ///        - LevelControl::StartupCurrentLevel has ID 0x4000
+        ///        - OnOff attributes GlobalSceneControl, OnTime, OffWaitTime, StartupOnOff have ID >= 0x4000
+        ///        - ColorControl, Thermostat and DoorLock have many attributes with high ID as well
+        ///
+        /// Use optionalAttributes only if its usage makes sense for the cluster. In many instances it does,
+        /// however it is not a generic rule. Usage of it must be double-checked as sufficient.
+        virtual ServerClusterRegistration & CreateRegistration(EndpointId endpointId, unsigned zeroBasedArrayIndex,
+                                                               uint32_t optionalAttributeBits, uint32_t featureMap) = 0;
+
+        // Find the previously created cluster on the given path
+        virtual ServerClusterInterface & FindRegistration(unsigned zeroBasedArrayIndex) = 0;
+
+        // Destroy/free the given registration
```

Comment:
Which one is it? Suggest "release the registration"

Reply:
Release felt vague. Realistically I have to call the constructor. I will update the comment to make this clear.